Cannot start in just services mode in Equinix Metal

**Describe the bug**

I tried to follow the documentation [here](https://kube-vip.io/docs/usage/equinix/) to create a manifest using the API. I want to run it just in services mode (no control plane).

I got confused because the section [Creating a manifest using the API](https://kube-vip.io/docs/usage/equinix/#creating-a-manifest-using-the-api) is mainly for control plane, when I want services; but the [section for services](https://kube-vip.io/docs/usage/equinix/#load-balancing-services-on-equinix-metal) mainly references the annotations (provided by the CCM) and using the existing CCM secret.

I had thought that I could use kube-vip's "Metal API knowledge" to connect directly to the API.

Further, it wasn't clear what `--vip $VIP` is supposed to do. Is that the VIP for control plane only (which is what I thought)?

I tried running it as follows:

```bash
ctr run --rm --net-host ${IMAGE} vip /kube-vip manifest daemonset \
  --interface lo \
  --metal \
  --metalKey $METAL_AUTH_TOKEN \
  --metalProjectID $METAL_PROJECT_ID \
  --services \
  --inCluster \
  --taint \
  --bgp > kube-vip.yaml
```

This works to create the yaml manifest, which then gives the following:

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  creationTimestamp: null
  labels:
    app.kubernetes.io/name: kube-vip-ds
    app.kubernetes.io/version: v0.7.2
  name: kube-vip-ds
  namespace: kube-system
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: kube-vip-ds
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/name: kube-vip-ds
        app.kubernetes.io/version: v0.7.2
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: node-role.kubernetes.io/master
                operator: Exists
            - matchExpressions:
              - key: node-role.kubernetes.io/control-plane
                operator: Exists
      containers:
      - args:
        - manager
        env:
        - name: vip_arp
          value: "false"
        - name: port
          value: "6443"
        - name: vip_interface
          value: lo
        - name: vip_cidr
          value: "32"
        - name: dns_mode
          value: first
        - name: svc_enable
          value: "true"
        - name: svc_leasename
          value: plndr-svcs-lock
        - name: vip_packet
          value: "true"
        - name: vip_packetproject
        - name: vip_packetprojectid
          value: ********
        - name: PACKET_AUTH_TOKEN
          value: ********
        - name: bgp_enable
          value: "true"
        - name: vip_address
        - name: prometheus_server
          value: :2112
        image: ghcr.io/kube-vip/kube-vip:v0.7.2
        imagePullPolicy: Always
        name: kube-vip
        resources: {}
        securityContext:
          capabilities:
            add:
            - NET_ADMIN
            - NET_RAW
      hostNetwork: true
      serviceAccountName: kube-vip
      tolerations:
      - effect: NoSchedule
        operator: Exists
      - effect: NoExecute
        operator: Exists
  updateStrategy: {}
status:
  currentNumberScheduled: 0
  desiredNumberScheduled: 0
  numberMisscheduled: 0
  numberReady: 0
```

When I applied it, it worked, deploying 3 pods in kube-system namespace.

When I then tried to deploy a service of `type:LoadBalancer`, it does odd things.

The service manifest is:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: jenkins-service
  namespace: jenkins
spec:
  selector:
    app: jenkins-server
  type: LoadBalancer
  ports:
    - port: 8080
      targetPort: 8080
```

The logs are as follows:

```
time="2024-04-15T17:42:44Z" level=info msg="Starting kube-vip.io [v0.7.2]"
time="2024-04-15T17:42:44Z" level=info msg="namespace [kube-system], Mode: [BGP], Features(s): Control Plane:[false], Services:[true]"
time="2024-04-15T17:42:44Z" level=warning msg="the status of the interface lo is unknown. Ensure your interface is ready to accept traffic, if so you can safely ignore this message"
time="2024-04-15T17:42:44Z" level=info msg="prometheus HTTP server started"
time="2024-04-15T17:42:44Z" level=info msg="Starting Kube-vip Manager with the BGP engine"
time="2024-04-15T17:42:44Z" level=info msg="Looking up the BGP configuration from Equinix Metal"
time="2024-04-15T17:42:45Z" level=info msg="Querying BGP settings for [controlplane-03]"
time="2024-04-15T17:42:45Z" level=info msg="Starting the BGP server to advertise VIP routes to BGP peers"
time="2024-04-15T17:42:45Z" level=info msg="Add a peer configuration" Key=169.254.255.1 Topic=Peer
time="2024-04-15T17:42:45Z" level=info msg="Add a peer configuration" Key=169.254.255.2 Topic=Peer
time="2024-04-15T17:42:45Z" level=info msg="(svcs) starting services watcher for all namespaces"
time="2024-04-15T17:42:45Z" level=info msg="[BGP] type:STATE  peer:{conf:{local_asn:65000  neighbor_address:\"169.254.255.1\"  peer_asn:65530}  state:{local_asn:65000  neighbor_address:\"169.254.255.1\"  peer_asn:65530  session_state:IDLE  router_id:\"<nil>\"}  transport:{local_address:\"<nil>\"}}"
time="2024-04-15T17:42:45Z" level=info msg="[BGP] type:STATE  peer:{conf:{local_asn:65000  neighbor_address:\"169.254.255.2\"  peer_asn:65530}  state:{local_asn:65000  neighbor_address:\"169.254.255.2\"  peer_asn:65530  session_state:IDLE  router_id:\"<nil>\"}  transport:{local_address:\"<nil>\"}}"
time="2024-04-15T17:42:45Z" level=info msg="[BGP] type:STATE  peer:{conf:{local_asn:65000  neighbor_address:\"169.254.255.1\"  peer_asn:65530}  state:{local_asn:65000  neighbor_address:\"169.254.255.1\"  peer_asn:65530  session_state:ACTIVE  router_id:\"<nil>\"}  transport:{local_address:\"<nil>\"}}"
time="2024-04-15T17:42:45Z" level=info msg="[BGP] type:STATE  peer:{conf:{local_asn:65000  neighbor_address:\"169.254.255.2\"  peer_asn:65530}  state:{local_asn:65000  neighbor_address:\"169.254.255.2\"  peer_asn:65530  session_state:ACTIVE  router_id:\"<nil>\"}  transport:{local_address:\"<nil>\"}}"
time="2024-04-15T17:42:51Z" level=info msg="[endpoints] watching for service [traefik] in namespace [kube-system]"
time="2024-04-15T17:42:51Z" level=info msg="(svcs) adding VIP [[10.65.154.3 10.65.154.5 139.178.80.171 139.178.80.191 139.178.80.201]] for [kube-system/traefik]"
E0415 17:42:59.258097       1 retrywatcher.go:129] "Watch failed" err="Get \"https://10.43.0.1:443/api/v1/services?watch=true\": dial tcp 10.43.0.1:443: connect: connection refused - error from a previous attempt: read tcp 10.65.154.3:48870->10.43.0.1:443: read: connection reset by peer"
time="2024-04-15T17:42:59Z" level=error msg="Failed to set Services: Get \"https://10.43.0.1:443/api/v1/namespaces/kube-system/services/traefik?timeout=10s\": dial tcp 10.43.0.1:443: connect: connection refused - error from a previous attempt: read tcp 10.65.154.3:48870->10.43.0.1:443: read: connection reset by peer"
time="2024-04-15T17:42:59Z" level=info msg="[LOADBALANCER] Stopping load balancers"
time="2024-04-15T17:42:59Z" level=info msg="[VIP] Releasing the Virtual IP [10.65.154.3]"
time="2024-04-15T17:42:59Z" level=info msg="[LOADBALANCER] Stopping load balancers"
time="2024-04-15T17:42:59Z" level=info msg="[VIP] Releasing the Virtual IP [10.65.154.5]"
time="2024-04-15T17:42:59Z" level=info msg="[LOADBALANCER] Stopping load balancers"
time="2024-04-15T17:42:59Z" level=info msg="[VIP] Releasing the Virtual IP [139.178.80.171]"
time="2024-04-15T17:42:59Z" level=info msg="[LOADBALANCER] Stopping load balancers"
time="2024-04-15T17:42:59Z" level=info msg="[VIP] Releasing the Virtual IP [139.178.80.191]"
time="2024-04-15T17:42:59Z" level=info msg="[LOADBALANCER] Stopping load balancers"
time="2024-04-15T17:42:59Z" level=info msg="[VIP] Releasing the Virtual IP [139.178.80.201]"
time="2024-04-15T17:42:59Z" level=info msg="Removed [02a6e484-e96a-4f9e-a808-31142c22209c] from manager, [0] advertised services remain"
time="2024-04-15T17:44:10Z" level=info msg="[endpoints] watching for service [jenkins-service] in namespace [jenkins]"
time="2024-04-15T17:44:10Z" level=info msg="(svcs) adding VIP [[139.178.80.201]] for [jenkins/jenkins-service]"
time="2024-04-15T17:44:10Z" level=info msg="[endpoints] deleted BGP host: 139.178.80.201/32, service: jenkins/jenkins-service"
time="2024-04-15T17:44:10Z" level=warning msg="(svcs) already found existing address [139.178.80.201] on adapter [lo]"
time="2024-04-15T17:44:10Z" level=error msg="Error updating Service jenkins/jenkins-service Status: Operation cannot be fulfilled on services \"jenkins-service\": the object has been modified; please apply your changes to the latest version and try again"
time="2024-04-15T17:44:10Z" level=info msg="[service] synchronised in 19ms"
time="2024-04-15T17:44:20Z" level=info msg="[endpoints] deleted BGP host: 139.178.80.201/32, service: jenkins/jenkins-service"
time="2024-04-15T17:44:20Z" level=info msg="[endpoints] deleted BGP host: 139.178.80.201/32, service: jenkins/jenkins-service"
time="2024-04-15T17:44:30Z" level=info msg="[endpoints] deleted BGP host: 139.178.80.201/32, service: jenkins/jenkins-service"
time="2024-04-15T17:44:40Z" level=info msg="[endpoints] deleted BGP host: 139.178.80.201/32, service: jenkins/jenkins-service"
time="2024-04-15T17:44:50Z" level=info msg="[endpoints] deleted BGP host: 139.178.80.201/32, service: jenkins/jenkins-service"
time="2024-04-15T17:45:00Z" level=info msg="[endpoints] deleted BGP host: 139.178.80.201/32, service: jenkins/jenkins-service"
time="2024-04-15T17:45:10Z" level=info msg="[endpoints] deleted BGP host: 139.178.80.201/32, service: jenkins/jenkins-service"
time="2024-04-15T17:45:20Z" level=info msg="[endpoints] deleted BGP host: 139.178.80.201/32, service: jenkins/jenkins-service"
time="2024-04-15T17:45:25Z" level=info msg="[endpoints] added BGP host: 139.178.80.201/32, service: jenkins/jenkins-service"
```

It looks like it is starting OK, creates active sessions to BGP peers. But then it seems to treat the actual nodes as VIPs:

```
time="2024-04-15T17:42:51Z" level=info msg="(svcs) adding VIP [[10.65.154.3 10.65.154.5 139.178.80.171 139.178.80.191 139.178.80.201]] for [kube-system/traefik]"
```

All of those addresses are actual (internal and external) IPs for the actual nodes, not a VIP.  Same for later on:

```
time="2024-04-15T17:42:59Z" level=info msg="[VIP] Releasing the Virtual IP [10.65.154.3]"
time="2024-04-15T17:42:59Z" level=info msg="[VIP] Releasing the Virtual IP [10.65.154.5]"
time="2024-04-15T17:42:59Z" level=info msg="[VIP] Releasing the Virtual IP [139.178.80.171]"
time="2024-04-15T17:42:59Z" level=info msg="[VIP] Releasing the Virtual IP [139.178.80.191]"
time="2024-04-15T17:42:59Z" level=info msg="[VIP] Releasing the Virtual IP [139.178.80.201]"
```

Those aren't Equinix EIPs, but the actual IPs on the various nodes. 

I also thought, perhaps I need to actually allocate an EIP and add it explicitly using `--vip` (even though I thought that is for control plane), but that didn't work any better. It actually seemed to kick out my ability to access the kube-apiserver at all for a bit.

What am I misunderstanding?

Update: my CCM was misconfigured. I got that right, but now I am seeing kube-vip with different errors.

I have 3 control plane and 3 worker nodes. On the control planes:

```
time="2024-04-16T14:44:39Z" level=info msg="Starting kube-vip.io [v0.7.2]"
time="2024-04-16T14:44:39Z" level=info msg="namespace [kube-system], Mode: [BGP], Features(s): Control Plane:[false], Services:[true]"
time="2024-04-16T14:44:39Z" level=warning msg="the status of the interface lo is unknown. Ensure your interface is ready to accept traffic, if so you can safely ignore this message"
time="2024-04-16T14:44:39Z" level=info msg="prometheus HTTP server started"
time="2024-04-16T14:44:39Z" level=info msg="Kube-Vip is waiting for annotation prefix [metal.equinix.com] to be present on this node"
time="2024-04-16T14:44:39Z" level=info msg="Starting Kube-vip Manager with the BGP engine"
time="2024-04-16T14:44:39Z" level=error msg="you must export PACKET_AUTH_TOKEN"
time="2024-04-16T14:44:39Z" level=info msg="Looking up the BGP configuration from Equinix Metal"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1e0 pc=0x16f7aa1]

goroutine 1 [running]:
github.com/kube-vip/kube-vip/pkg/equinixmetal.findProject({0x0, 0x0}, 0x0)
	/src/pkg/equinixmetal/utils.go:14 +0xa1
github.com/kube-vip/kube-vip/pkg/equinixmetal.BGPLookup(0x0, 0x32d9c00)
	/src/pkg/equinixmetal/bgp.go:16 +0x56
github.com/kube-vip/kube-vip/pkg/manager.(*Manager).startBGP(0xc0001ea930)
	/src/pkg/manager/manager_bgp.go:46 +0x211
github.com/kube-vip/kube-vip/pkg/manager.(*Manager).Start(0xc0001ea930)
	/src/pkg/manager/manager.go:167 +0x2e5
github.com/kube-vip/kube-vip/cmd.init.func12(0x32bc9c0, {0x1fa6201?, 0x4?, 0x1fa6205?})
	/src/cmd/kube-vip.go:362 +0xa9b
github.com/spf13/cobra.(*Command).execute(0x32bc9c0, {0x333a840, 0x0, 0x0})
	/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:987 +0xab1
github.com/spf13/cobra.(*Command).ExecuteC(0x32bbe40)
	/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039
github.com/kube-vip/kube-vip/cmd.Execute()
	/src/cmd/kube-vip.go:173 +0x1a
main.main()
	/src/main.go:17 +0x85
```

On the workers:

```
Error from server: Get "https://10.65.154.11:10250/containerLogs/kube-system/kube-vip-ds-n7gb4/kube-vip": tls: failed to verify certificate: x509: certificate is valid for 127.0.0.1, 139.178.80.201, 2604:1380:4643:2800::9, not 10.65.154.11
```

Almost identical message repeats itself on other words. FWIW, the `.201` is the public address of the worker, the 10.65.154.11` is the private address.

--- a/pkg/manager/manager_bgp.go
+++ b/pkg/manager/manager_bgp.go
@@ -27,7 +27,7 @@ func (sm *Manager) startBGP() error {
 		if sm.config.ProviderConfig != "" {
 			key, project, err := equinixmetal.GetPacketConfig(sm.config.ProviderConfig)
 			if err != nil {
-				log.Error(err)
+				return err
 			} else {
 				// Set the environment variable with the key for the project
 				os.Setenv("PACKET_AUTH_TOKEN", key)
@@ -37,7 +37,7 @@ func (sm *Manager) startBGP() error {
 		}
 		packetClient, err = packngo.NewClient()
 		if err != nil {
-			log.Error(err)
+			return err
 		}
 
 		// We're using Equinix Metal with BGP, populate the Peer information from the API
@@ -45,7 +45,7 @@ func (sm *Manager) startBGP() error {
 			log.Infoln("Looking up the BGP configuration from Equinix Metal")
 			err = equinixmetal.BGPLookup(packetClient, sm.config)
 			if err != nil {
-				log.Error(err)
+				return err
 			}
 		}
 	}

--- a/pkg/manager/manager_bgp.go
+++ b/pkg/manager/manager_bgp.go
@@ -28,12 +28,12 @@ func (sm *Manager) startBGP() error {
 			key, project, err := equinixmetal.GetPacketConfig(sm.config.ProviderConfig)
 			if err != nil {
 				return err
-			} else {
-				// Set the environment variable with the key for the project
-				os.Setenv("PACKET_AUTH_TOKEN", key)
-				// Update the configuration with the project key
-				sm.config.MetalProjectID = project
 			}
+			// Set the environment variable with the key for the project
+			os.Setenv("PACKET_AUTH_TOKEN", key)
+			// Update the configuration with the project key
+			sm.config.MetalProjectID = project
+
 		}
 		packetClient, err = packngo.NewClient()
 		if err != nil {

